src/charm.py

```diff
@@ -110,6 +188,47 @@ def __init__(self, *args):
         self.framework.observe(
             self.on.validate_configuration_action, self._on_validate_configuration
         )
+        self.framework.observe(self.resources_patch.on.patch_failed, self._on_k8s_patch_failed)
+
+    @classmethod
+    def _limits_to_requests(cls, resource_limits: ResourceSpecDict):
+        """Calculate "requests" from a "limits" dict.
+
+        By default, we set the "requests" portion of the resource limits to a sensible value, while
+        keeping the "limits" portion unspecified.
+        If the user specifies a "limits" (via config options), then the "requests" portion is
+        calculated to maintain a monotone function, as follows:
+        - Calculate the adjusted limit by multiplying the user value by 0.8.
+        - If the adjusted value is larger than the default "requests", use the adjusted value for
```

Comment:
I appreciate the effort this detailed docstring. However would this be clearer if said into code ? And instead would it make more sense to make clear in the docstring what is the intended purpose of this logic -- for example to ensure that the "requests" are not more than the "limits"  and that this constraining uses a value of limits that is scaled down by a multiplier.

Also isn't this functionality more rightly placed in the context of the kubernetes resource patch library along with the other methods above. ?

Reply:
Fixed in next commit, lemme know what you think!